include/image-commands.mk

```diff
@@ -217,6 +217,96 @@ define Build/append-kernel
 	dd if=$(IMAGE_KERNEL) >> $@
 endef
 
+# \n = $(newline), so that a mutli-line variable can be used
+# in shell through substituting $(newline) with '\n'
+# https://stackoverflow.com/a/5887751
+define newline
+
+
+endef
+
+define Build/append-raw-dtb-to-elf-linker-script
+OUTPUT_ARCH(ARCH_LD)
+SECTIONS {
+	/*
+	 * Preprocess the load address
+	 * -Ttext $ADDR is brittle:
+	 * https://lkml.kernel.org/lkml/20200413153453.zi4jvu3c4ul23e23@google.com/
+	 */
+	. = ZLOADADDR;
+	.text : {
+		startup = .;
+		*(.text)
+		*(.text.*)
+		*(.data)
+		*(.data.*)
+	}
+}
+endef
+
+# Replaces kernel-bin
+# requires ELF kernel
+# uses $@: target_rule_name example: /**/build_dir/target-*/linux-*/tmp/openwrt-*{,-initramfs}-kernel.bin
+# uses $<: first_prerequisite_name example: /**/build_dir/target-*/linux-*/$KERNEL_NAME{,-initramfs}
+# process:
+# copy the $KERNEL_NAME to $KDIR/tmp
+# convert to bytecode
+# append DTB
+# preprocess linker script https://elixir.bootlin.com/linux/v5.4.66/source/scripts/Makefile.build#L353
+# convert back to ELF (which requires data from the original ELF)
+# - ELF type example: elf32-tradlittlemips
+# - load address example: 0x80931000
+define Build/append-raw-dtb-to-elf
+	$(eval $@ : KERNEL_ELF_INPUT := $<)
+	$(eval $@ : KERNEL_OUTPUT := $@)
+	$(eval $@ : KERNEL_TEMP_UNLINKED := $(KERNEL_OUTPUT)_unlinked)
+	$(eval $@ : KERNEL_TEMP_BC := $(KERNEL_OUTPUT)_binary)
+	$(eval $@ : LINKER_SCRIPT := $(KERNEL_OUTPUT).lds)
+	$(eval $@ : LINKER_SCRIPT_UNPROCESSED := $(LINKER_SCRIPT).S)
+
+	$(eval $@ : ZLOADADDR := $(filter-out Entry point address:, \
+				   $(shell readelf -h $(KERNEL_ELF_INPUT) | \
+				     grep Entry)))
+	$(eval $@ : ELF_FORMAT := $(filter-out $(KERNEL_ELF_INPUT): file format, \
+				    $(shell $(TARGET_CROSS)objdump -h $(KERNEL_ELF_INPUT) | \
+				      grep 'file format' )))
+
+	$(eval $@ : ARCH_LD := $(subst mipsel,mips,$(ARCH)))
+
+	$(eval $@ : TEMP_FILENAMES := $(KERNEL_TEMP_UNLINKED) \
+				      $(KERNEL_TEMP_BC) \
+				      $(LINKER_SCRIPT) \
+				      $(LINKER_SCRIPT_UNPROCESSED))
+
+	$(info KERNEL_OUTPUT=$(KERNEL_OUTPUT))
+	$(info KERNEL_ELF_INPUT=$(KERNEL_ELF_INPUT))
+	$(info ZLOADADDR=$(ZLOADADDR))
+	$(info ELF_FORMAT=$(ELF_FORMAT))
+	$(info ARCH_LD=$(ARCH_LD))
+
+	rm -f $(KERNEL_OUTPUT)
+	rm -f $(TEMP_FILENAMES)
+
+	$(TARGET_CROSS)objcopy -O binary \
+			       -R .reginfo -R .note -R .comment -R .mdebug -R .MIPS.abiflags -S \
```

Comment:
It seems that we don't need all these -R flags according to include/kernel-defaults.mk?
I was copying them from lzma-loader Makefile. Just drop them if it's not needed.